File: 0042-trapping-rain-water/0042-trapping-rain-water.kt
class Solution {
    fun trap(height: IntArray): Int {
    var left = 0
    var right = height.size - 1
    var leftMax = 0
    var rightMax = 0
    var totalWater = 0

    while (left < right) {
        if (height[left] < height[right]) {
            if (height[left] > leftMax) {
                leftMax = height[left]
            } else {
                totalWater += leftMax - height[left]
            }
            left++
        } else {
            if (height[right] > rightMax) {
                rightMax = height[right]
            } else {
                totalWater += rightMax - height[right]
            }
            right--
        }
    }

    return totalWater  
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.go
func trap(height []int) int {
    left := 0
    right := len(height) - 1
    left_max := 0
    right_max := 0
    total_water := 0

    for left < right {
        if height[left] < height[right] {
            if height[left] > left_max {
                left_max = height[left]
            } else {
                total_water += left_max - height[left]
            }
            left++;
        } else {
            if height[right] > right_max {
                right_max = height[right]
            } else {
                total_water += right_max - height[right]
            }
            right--
        }
    }
    return total_water
}